osu.Game/PerformFromMenuRunner.cs

```diff
@@ -0,0 +1,145 @@
+    {
+        private readonly Action<IScreen> finalAction;
+        private readonly IEnumerable<Type> validScreens;
+        private readonly Func<IScreen> getCurrentScreen;
+
+        [Resolved]
+        private NotificationOverlay notifications { get; set; }
+
+        [Resolved]
+        private DialogOverlay dialogOverlay { get; set; }
+
+        [Resolved]
+        private IBindable<WorkingBeatmap> beatmap { get; set; }
+
+        [Resolved(canBeNull: true)]
+        private OsuGame game { get; set; }
+
+        private readonly ScheduledDelegate task;
+
+        private PopupDialog lastEncounteredDialog;
+        private IScreen lastEncounteredDialogScreen;
+
+        /// <summary>
+        /// Perform an action only after returning to a specific screen as indicated by <paramref name="validScreens"/>.
+        /// Eagerly tries to exit the current screen until it succeeds.
+        /// </summary>
+        /// <param name="finalAction">The action to perform once we are in the correct state.</param>
+        /// <param name="validScreens">An optional collection of valid screen types. If any of these screens are already current we can perform the action immediately, else the first valid parent will be made current before performing the action. <see cref="MainMenu"/> is used if not specified.</param>
+        /// <param name="getCurrentScreen">A function to retrieve the currently displayed game screen.</param>
+        public PerformFromMenuRunner(Action<IScreen> finalAction, IEnumerable<Type> validScreens, Func<IScreen> getCurrentScreen)
+        {
+            validScreens ??= Enumerable.Empty<Type>();
+            validScreens = validScreens.Append(typeof(MainMenu));
+
+            this.finalAction = finalAction;
+            this.validScreens = validScreens;
+            this.getCurrentScreen = getCurrentScreen;
+
+            Scheduler.Add(task = new ScheduledDelegate(checkCanComplete, 0, 200));
+        }
+
+        /// <summary>
+        /// Cancel this runner from running.
+        /// </summary>
+        public void Cancel()
+        {
+            task.Cancel();
+            Expire();
+        }
+
+        private void checkCanComplete()
+        {
+            // find closest valid target
+            IScreen current = getCurrentScreen();
+
+            // a dialog may be blocking the execution for now.
+            if (checkForDialog(current)) return;
+
+            // we may already be at the target screen type.
+            if (validScreens.Contains(getCurrentScreen().GetType()) && !beatmap.Disabled)
+            {
+                complete();
+                return;
+            }
+
+            game.CloseAllOverlays(false);
+
+            while (current != null)
+            {
+                if (validScreens.Contains(current.GetType()))
+                {
+                    current.MakeCurrent();
+                    break;
+                }
+
+                current = current.GetParentScreen();
+            }
```

Comment:
This seems to also have been a theoretical problem earlier before, but what happens if none of the target screens to exit to is in the stack? If I'm reading correctly, looks like it'll spin every 200ms until canceled by the next perform.

Reply:
MainMenu is always added and should always be in the stack (that was the intention).

Reply:
Considered potentially checking the hierarchy for the presence of `MainMenu` and aborting if not there, but I don't think that's the behaviour we want. This `PresentFromMainMenu` may be called during the startup process where we are not at the menu yet, and it should still work in such cases.